### PR TITLE
Only serve compatible APIs for syncTarget

### PR DESCRIPTION
**Pull request by qiujian16**
Signed-off-by: Jian Qiu <jqiu@redhat.com>

## Related issue(s)

Fixes https://github.com/kcp-dev/kcp/issues/1779

followup of this PR: https://github.com/kcp-dev/kcp/pull/1624

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
@qiujian16 it would really help reviewers if you could link to an epic/a specific item/goal for the PR, so I have some context to ask "does this achieve the goal we have" :)

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 1]
```diff
@@ -176,6 +152,60 @@ func gvrString(gvr schema.GroupVersionResource) string {
 	return fmt.Sprintf("%s.%s.%s", gvr.Resource, gvr.Version, group)
 }
 
+func (c *APIReconciler) getAllCompatibleResourceSchemas(syncTarget *workloadv1alpha1.SyncTarget) (map[schema.GroupResource]*apisv1alpha1.APIResourceSchema, map[schema.GroupResource]string, error) {
```
Comment (stevekuznetsov) [inline pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:155]: What does `compatible` mean here? As I read it, this function accrues all schemas from all referenced exports that are already synced?

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 2]
```diff
@@ -176,6 +152,60 @@ func gvrString(gvr schema.GroupVersionResource) string {
...
+	apiExportKeys := getExportKeys(syncTarget)
+	apiResourceSchemas := map[schema.GroupResource]*apisv1alpha1.APIResourceSchema{}
+
+	identityHash := map[schema.GroupResource]string{}
```
Comment (stevekuznetsov) [inline pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:159]:
```suggestion
	identityHashByGroupResource := map[schema.GroupResource]string{}
```

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 3]
```diff
@@ -28,87 +28,63 @@ import (
...
 	oldSet := c.apiSets[apiDomainKey]
 	c.mutex.RUnlock()
 
-	// collect APIResourceSchemas
-	schemaIdentity := map[string]string{}
-	apiResourceSchemas := make([]*apisv1alpha1.APIResourceSchema, 0, len(apiExport.Spec.LatestResourceSchemas)+len(syncerSchemas))
-	for _, schema := range syncerSchemas {
-		shallow := *schema
+	// collect APIResourceSchemas by syncTarget.
+	apiResourceSchemas, schemaIdentity, err := c.getAllCompatibleResourceSchemas(syncTarget)
```
Comment (stevekuznetsov) [inline pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:56]:
```suggestion
	apiResourceSchemas, schemaIdentites, err := c.getAllCompatibleResourceSchemas(syncTarget)
```

**Review by stevekuznetsov — COMMENTED** [review]
It's hard to say if the logic here is correct - I would like to understand better what it means to be "compatible" and perhaps more context for this PR, and/or tests to show the flows.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go (hunk 1 above)
Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:155]: this is the followup of this one https://github.com/kcp-dev/kcp/pull/1624 which check if synced resource is compatible with schema defined in exports.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_controller.go [hunk 4]
```diff
@@ -81,30 +80,29 @@ func NewAPIReconciler(
...
-		},
-		DeleteFunc: func(obj interface{}) {
-			c.enqueueSyncTarget(obj)
+		AddFunc: func(obj interface{}) { c.enqueueSyncTarget(obj, "") },
+		UpdateFunc: func(old, obj interface{}) {
+			oldCluster := old.(*workloadv1alpha1.SyncTarget)
+			newCluster := obj.(*workloadv1alpha1.SyncTarget)
+
+			// only enqueue when syncedResource is changed.
+			if !equality.Semantic.DeepEqual(oldCluster.Status.SyncedResources, newCluster.Status.SyncedResources) {
```
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_controller.go:101]: the syncTarget updated quite frequent (with lastHeartBeat in status) and this controller does not care.

**Comment by qiujian16** [comment]
> @qiujian16 it would really help reviewers if you could link to an epic/a specific item/goal for the PR, so I have some context to ask "does this achieve the goal we have" :)

linked to the related issue.

**Comment by qiujian16** [comment]
> It's hard to say if the logic here is correct - I would like to understand better what it means to be "compatible" and perhaps more context for this PR, and/or tests to show the flows.

yeah, still working on the e2e tests.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 5]
```diff
@@ -176,6 +152,60 @@ func gvrString(gvr schema.GroupVersionResource) string {
...
+				Resource: syncedResource.Resource,
+			}] = syncedResource.IdentityHash
+		}
+	}
+
+	var errs []error
+	for _, apiExportKey := range apiExportKeys {
+		apiExport, err := c.apiExportLister.Get(apiExportKey)
+		if apierrors.IsNotFound(err) {
+			klog.V(3).Infof("APIExport %s not found", apiExportKey)
```
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:175]: do you mean using `klog.FromContext(ctx)`? Try to find an example.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go (hunk 1 above)
Comment (stevekuznetsov) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:155]: If you could define compatibility - perhaps in a comment for this method - that would help me wrap my head around this logic

Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go]: comments added. PTAL

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_indexes.go [hunk 6]
```diff
@@ -21,15 +21,54 @@ import (
...
+	if len(synctarget.Spec.SupportedAPIExports) == 0 {
+		return []string{clusters.ToClusterAwareKey(lcluster, reconcilerapiexport.TemporaryComputeServiceExportName)}
+	}
+
+	var keys []string
+	for _, export := range synctarget.Spec.SupportedAPIExports {
+		if export.Workspace == nil {
+			continue
+		}
+		if len(export.Workspace.Path) == 0 {
```
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_indexes.go:66]: we only default the name, but not the path.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go]: updated to use contextual logging

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_indexes.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_indexes.go:66]: The Path is defaulted in admission for references in `APIBinding`s, but not for `SyncTarget`s afaict from the code.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 7]
```diff
@@ -28,116 +28,95 @@ import (
...
-		defer c.mutex.Unlock()
-		klog.V(2).Infof("Deleting APIs for API domain key %s", apiDomainKey)
-		delete(c.apiSets, apiDomainKey)
-		return nil
-	}
-
-	if apiExport.ObjectMeta.Name != apiexport.TemporaryComputeServiceExportName {
-		// this is not something we're handling in this controller
+func (c *APIReconciler) reconcile(ctx context.Context, apiDomainKey dynamiccontext.APIDomainKey, syncTarget *workloadv1alpha1.SyncTarget) error {
+	if len(syncTarget.Status.SyncedResources) == 0 {
```
Comment (davidfestal) [inline pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:47]: Do we really want to do that here ?
What about the built-in resources (like coinfig maps, secrets, ...)
It seems to me that an APIefinitionSet with only built-in resources might make sense (even though not very useful ;-) )

Comment (qiujian16) [reply pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go]: right, removed this part.

File: pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go [hunk 8]
```diff
@@ -176,6 +150,60 @@ func gvrString(gvr schema.GroupVersionResource) string {
 	return fmt.Sprintf("%s.%s.%s", gvr.Resource, gvr.Version, group)
 }
 
+// getAllAcceptedResourceSchemas return all resourceSchemas from APIExports defined in ths syncTarget filtered by the status.syncedResource
```
Comment (sttts) [inline pkg/virtual/syncer/controllers/apireconciler/syncer_apireconciler_reconcile.go:153]:
```suggestion
// getAllAcceptedResourceSchemas return all resourceSchemas from APIExports defined in this syncTarget filtered by the status.syncedResource
```

**Comment by sttts** [comment]
Needs a rebase.

Sgtm.

**Pull request merged**